geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoFeatureSerializer.scala

```diff
@@ -24,51 +23,97 @@ import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 import scala.collection.JavaConversions._
 
 class KryoFeatureSerializer(sft: SimpleFeatureType, val options: Set[SerializationOption] = Set.empty)
-    extends SimpleFeatureSerializer with SimpleFeatureDeserializer with LazyLogging {
+    extends KryoSimpleFeatureSerialization with KryoSimpleFeatureDeserialization {
+  override private [kryo] def serializeSft = sft
+  override private [kryo] def deserializeSft = sft
+}
 
-  import KryoFeatureSerializer._
+/**
+  * @param original the simple feature type that will be serialized
+  * @param projected the simple feature type to project to when serializing
+  */
+class ProjectingKryoFeatureSerializer(original: SimpleFeatureType,
+                                      projected: SimpleFeatureType,
+                                      val options: Set[SerializationOption] = Set.empty)
+    extends SimpleFeatureSerializer with KryoSimpleFeatureDeserialization {
 
-  protected[kryo] val cacheKey = CacheKeyGenerator.cacheKeyForSFT(sft)
-  protected[kryo] val numAttributes = sft.getAttributeCount
+  import KryoFeatureSerializer._
 
-  protected[kryo] val writers = getWriters(cacheKey, sft)
-  protected[kryo] val readers = getReaders(cacheKey, sft)
+  require(!options.withUserData, "User data serialization not supported")
 
-  private lazy val legacySerializer = serialization.KryoFeatureSerializer(sft, sft, options)
-  private lazy val lazyUserData: (Input) => jMap[AnyRef, AnyRef] = if (options.withUserData) {
-    (input) => kryoReader.readGenericMap(VERSION)(input)
-  } else {
-    (input) => new jHashMap[AnyRef, AnyRef]
-  }
+  override private [kryo] def deserializeSft = projected
 
-  def getReusableFeature: KryoBufferSimpleFeature = new KryoBufferSimpleFeature(sft, readers, lazyUserData)
+  private val cacheKey = CacheKeyGenerator.cacheKeyForSFT(projected)
+  private val numAttributes = projected.getAttributeCount
+  private val writers = getWriters(cacheKey, projected)
+  private val mappings = Array.ofDim[Int](numAttributes)
 
-  override def serialize(sf: SimpleFeature): Array[Byte] = doWrite(sf)
-  override def deserialize(bytes: Array[Byte]): SimpleFeature = doRead(bytes)
+  projected.getAttributeDescriptors.zipWithIndex.foreach { case (d, i) =>
+    mappings(i) = original.indexOf(d.getLocalName)
+  }
 
-  override def serialize(i: Int, value: AnyRef): Array[Byte] = {
+  override def serialize(sf: SimpleFeature): Array[Byte] = {
+    val offsets = getOffsets(cacheKey, numAttributes)
     val output = getOutput()
-    writers(i)(output, value)
+    output.writeInt(VERSION, true)
+    output.setPosition(5) // leave 4 bytes to write the offsets
+    if (!options.withoutId) {
+      output.writeString(sf.getID)  // TODO optimize for uuids?
```

Comment:
this is a left over TODO from a long time ago that just got moved a bit - basically if we know that IDs are UUIDs we could encode them as 2 longs instead of a string, which would be smaller on disk.